### ISSUE TITLE
ES6, node and electron support

### ISSUE THREAD

**Issue by mikolalysenko** [stackgl/glslify#80]
glslify needs some maintenance to keep up with changes in the broader JavaScript ecosystem.  It is currently quite onerous to set up in electron or node environments and it also has issues with ES6 syntax.

In the next version we should take advantage of things like tagged template strings for writing shaders inline.

In order to support node and electron out of the box though we'll need to break the interface in 3 major ways:  

1.  Because we can't figure out the `__dirname` of the module which is invoking glslify as a function, we'll need to modify module.exports to return a constructor object which then creates the new properly configured glslify instance.  
1. Invoking glslify should return a string immediately, so we'll have to modify glslify-deps and glslify-resolve to be synchronous.
1.  Calling glslify directly now invokes glslify so the transform interface will have to be split out into a separate module or moved to a different entry point (`glslify/transform`, `glslify-transform`?)

**Comment by hughsk** [comment]
👋

What would the `__dirname` fix look like? I know @yoshuawuyts and @ahdinosaur had a [solution for sheetify in Node](https://github.com/stackcss/sheetify/blob/ec2a491fed4ee1632c3b9ca9243d40ba75789f50/index.js#L29) that worked around the problem without changing API, but I'm not fully aware of the tradeoffs.

Would be great for us to avoid making breaking API changes, or at least make the transition straightforward, to minimize the impact on pixi.js and others :)

If moving to `glslify/transform` or `glslify-transform`, is there a way to trigger an error/warning when someone uses `glslify` as a browserify transform directly?

**Comment by substack** [comment]
My intuition around the sheetify stacktrace hack is that it's a fragile solution which could be unsuitable for unforseen new environments. I think a new API where __dirname is passed in explicitly would not be very hard to transition to:

``` js
var glx = require('glslify')(__dirname)
```

and then you can call `glx.load(file)`, `glx.compile(src)`, or use `glx` as a tagged template string function:

``` js
console.log(glx`
  #pragma glslify: noise = require('glsl-noise/simplex/3d')
  precision mediump float;
  varying vec3 vpos;
  void main () {
    gl_FragColor = vec4(noise(vpos*25.0),1);
  }
`)
```

I have some code working for this interface that works in node/electron natively and also in the browser with a browserify transform.

To port existing code to this new proposed system:

* `require('glslify')` becomes `require('glslify')(__dirname)`
* `glslify(file)` becomes `glslify.file(file)`

**Comment by substack** [comment]
Branch with current progress: https://github.com/substack/glslify/tree/v6

**Comment by mikolalysenko** [comment]
That sheetify hack seems like an acceptable solution to me.  While it may be fragile, I think not breaking the API is a higher priority.  So :+1: for that hack.

Regarding template strings vs file names:  I think that template strings seem like a better default than taking a file as an argument.

@hughsk do you have any thoughts on how we could support this?  One way I was thinking about was to do dynamic dispatch on the arguments and check if the input is an array of strings (or input to a glslify string) and if so then running the template string version, otherwise doing a file look up.

**Comment by substack** [comment]
These features have been published in 6.0.0.
